src/System.Private.ServiceModel/src/Internals/System/Runtime/AsyncLock.cs

```diff
@@ -18,101 +19,130 @@ internal class AsyncLock
 #endif
         private readonly SemaphoreSlim _semaphore;
         private readonly SafeSemaphoreRelease _semaphoreRelease;
-        private AsyncLocal<bool> _lockTaken;
+        private static AsyncLocal<object> s_heldLocks = new AsyncLocal<object>(LockTakenValueChanged);
 
         public AsyncLock()
         {
             _semaphore = new SemaphoreSlim(1);
             _semaphoreRelease = new SafeSemaphoreRelease(this);
-            _lockTaken = new AsyncLocal<bool>(LockTakenValueChanged);
-            _lockTaken.Value = false;
         }
 
-        private void LockTakenValueChanged(AsyncLocalValueChangedArgs<bool> obj)
+        private static void LockTakenValueChanged(AsyncLocalValueChangedArgs<object> obj)
         {
             // Without this fixup, when completing the call to await TakeLockAsync there is
-            // a switch of Context and _localTaken will be reset to false. This is because
+            // a switch of Context and _heldLocks will be reset to null. This is because
             // of leaving the task.
 
             if (obj.ThreadContextChanged)
             {
-                _lockTaken.Value = obj.PreviousValue;
+                s_heldLocks.Value = obj.PreviousValue;
             }
```

Comment:
I'm surprised this handler doesn't lead to other problems, like a lock showing as being held even when it's not.  If I'm understanding this handler correctly, it's basically undoing the undo?  That would suggest if you had code like:
```C#
public async Foo()
{
    await TakeLock();
    await Something();
    ReleaseLock();
}
...
Task t = Foo();
// is lock held here?
```
that if TakeLock completed synchronously and then Something completed asynchronously, the code after `Task t = Foo()` would see the lock as being held?